crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs

```diff
@@ -17,22 +17,12 @@ where
 {
     ctx_b.validate_message_signer(&msg.signer)?;
 
-    // An IBC connection running on the local (host) chain should exist.
-    if msg.connection_hops_on_b.len() != 1 {
-        return Err(ChannelError::InvalidConnectionHopsLength {
-            expected: 1,
-            actual: msg.connection_hops_on_b.len(),
-        })
-        .map_err(ContextError::ChannelError);
-    }
+    // Note: Verify that the provided message contains only one connection end
+    // has been done during the conversion from the proto to the domain type.
```

Comment:
I find this harder to understand... The checks are scattered throughout the codebase, which makes it hard to verify that any one check is present when verifying the correctness of a handler.

Reply:
Yeah how I initially did the `validate()` is that it would contain *all* the logic, and nowhere else. Note that if the domain types are private, then we know that we'll never need to e.g "just call `validate_basic()`"; all our validation calls are within the `validation()` functions. So I prefer to just put all the validation calls directly in those functions, and none in the types directly (i.e. `try_from()`, `new()`, etc). Basically, every check should be traceable from a call that I can see in a `validate()`; so an explicit call to `validate_basic()` is OK, whereas a check in a `TryFrom` isn't.

WDYT?

Reply:
See if you like it? [cb71f2c](https://github.com/cosmos/ibc-rs/pull/650/commits/cb71f2c2c1ae0706b1440fd17c0c46fc964a3853)
It's been a bit tricky to avoid code duplication around connection_hops verification and meanwhile expose that method anywhere needed. Some places it's called on `ChannelEnd`, others on Msg types like `MsgChannelOpenTry` and `MsgChannelOpenInit`